File: windows/Max.Core/Channels/SlackChannel.cs
using System.Net.Http.Headers;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json.Nodes;

namespace Max.Core;

/// <summary>
/// Slack channel over Socket Mode: apps.connections.open → WebSocket → events_api
/// envelopes, replying with chat.postMessage. Needs an app-level token (xapp-, with
/// connections:write) and a bot token (xoxb-). Allowlist = your Slack member ID (U…).
/// </summary>
public sealed class SlackChannel
{
    private static readonly HttpClient Http = new();
    private readonly ISecretStore _secrets;
    private readonly ChannelRouter _router;
    private CancellationTokenSource? _cts;

    public SlackChannel(ISecretStore secrets, ChannelRouter router) { _secrets = secrets; _router = router; }

    public void Start()
    {
        var config = MaxConfig.Load();
        var appToken = _secrets.Get("slack-app");
        var botToken = _secrets.Get("slack-bot");
        if (!config.SlackEnabled || string.IsNullOrEmpty(appToken) || string.IsNullOrEmpty(botToken)) return;
        if (_cts != null) return;
        _cts = new CancellationTokenSource();
        _ = RunAsync(appToken!, botToken!, _cts.Token);
    }

    public void Stop() { _cts?.Cancel(); _cts = null; }

    private async Task RunAsync(string appToken, string botToken, CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try { await ConnectOnceAsync(appToken, botToken, ct); }
            catch (OperationCanceledException) { return; }
            catch (Exception ex) { ChannelLog.Write($"slack: {ex.Message}"); }
            try { await Task.Delay(5000, ct); } catch { return; }
        }
    }

    private async Task ConnectOnceAsync(string appToken, string botToken, CancellationToken ct)
    {
        var wssUrl = await OpenConnectionAsync(appToken, ct);
        if (wssUrl is null) { try { await Task.Delay(10000, ct); } catch { } return; }

        using var ws = new ClientWebSocket();
        await ws.ConnectAsync(new Uri(wssUrl), ct);
        ChannelLog.Write("slack: connected (socket mode)");

        while (ws.State == WebSocketState.Open && !ct.IsCancellationRequested)
        {
            var raw = await ReceiveAsync(ws, ct);
            if (raw is null) break;
            JsonObject msg;
            try { msg = JsonNode.Parse(raw)!.AsObject(); } catch { continue; }

            var type = (string?)msg["type"];
            if (type == "disconnect") break;
            if (type != "events_api") continue;

            // Acknowledge the envelope immediately.
            if ((string?)msg["envelope_id"] is string envId)
            {
                var ackBytes = Encoding.UTF8.GetBytes(new JsonObject { ["envelope_id"] = envId }.ToJsonString());
                await ws.SendAsync(ackBytes, WebSocketMessageType.Text, true, ct);
            }

            var ev = msg["payload"]?["event"]?.AsObject();
            if (ev is null || (string?)ev["type"] != "message") continue;
            if (ev["bot_id"] != null || ev["subtype"] != null) continue; // skip bots/edits/joins

            var userId = (string?)ev["user"] ?? "";
            var text = (string?)ev["text"] ?? "";
            var channel = (string?)ev["channel"] ?? "";
            if (string.IsNullOrWhiteSpace(text) || channel.Length == 0) continue;

            var allow = MaxConfig.Load().SlackAllowlist;
            if (!allow.Contains(userId))
            {
                ChannelLog.Write($"slack: ignored message from {userId} — not in allowlist");
                continue;
            }

            ChannelLog.Write($"slack: handling message from {userId}");
            await _router.HandleAsync($"slack-{channel}", text.Trim(),
                reply => PostMessageAsync(channel, reply, botToken), ct);
        }
    }

    private static async Task<string?> OpenConnectionAsync(string appToken, CancellationToken ct)
    {
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Post, "https://slack.com/api/apps.connections.open");
            req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", appToken);
            using var resp = await Http.SendAsync(req, ct);
            var obj = JsonNode.Parse(await resp.Content.ReadAsStringAsync(ct))?.AsObject();
            if (obj?["ok"]?.GetValue<bool>() == true) return (string?)obj["url"];
            ChannelLog.Write($"slack: connections.open error {obj?["error"]}");
            return null;
        }
        catch (Exception ex) { ChannelLog.Write($"slack: open error {ex.Message}"); return null; }
    }

    private static async Task PostMessageAsync(string channel, string text, string botToken)
    {
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Post, "https://slack.com/api/chat.postMessage");
            req.Headers.Authorization = new AuthenticationHeaderValue("Bearer", botToken);
            var body = new JsonObject { ["channel"] = channel, ["text"] = text };
            req.Content = new StringContent(body.ToJsonString(), Encoding.UTF8, "application/json");
            await Http.SendAsync(req);
        }
        catch (Exception ex) { ChannelLog.Write($"slack: send error {ex.Message}"); }
    }

    private static async Task<string?> ReceiveAsync(ClientWebSocket ws, CancellationToken ct)
    {
        var buffer = new byte[16 * 1024];
        using var ms = new MemoryStream();
        WebSocketReceiveResult result;
        do
        {
            try { result = await ws.ReceiveAsync(buffer, ct); }
            catch { return null; }
            if (result.MessageType == WebSocketMessageType.Close) return null;
            ms.Write(buffer, 0, result.Count);
        } while (!result.EndOfMessage);
        return Encoding.UTF8.GetString(ms.ToArray());
    }
}

File: Sources/AskMax/Core/Channels/SlackChannel.swift
import Foundation

/// Slack bot via Socket Mode (WebSocket) — no public webhook required, which is
/// what makes it work for a local Mac app. Flow: open a socket URL with the
/// app-level token (xapp-…), receive event envelopes, ACK each, handle message
/// events, and reply with chat.postMessage using the bot token (xoxb-…).
@MainActor
final class SlackChannel {
    static let shared = SlackChannel()

    private var task: URLSessionWebSocketTask?
    private var running = false
    private var appToken = ""
    private var botToken = ""
    private var selfUserId = "" // bot's own user id, to ignore its own messages

    private init() {}

    func startIfEnabled() {
        let config = MaxConfig.load()
        guard config.slackEnabled,
              let app = SecretStore.secret(forKey: "slack-app"), !app.isEmpty,
              let bot = SecretStore.secret(forKey: "slack-bot"), !bot.isEmpty else {
            stop(); return
        }
        guard !running else { return }
        appToken = app
        botToken = bot
        running = true
        openSocket()
    }

    func stop() {
        running = false
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }

    private func openSocket() {
        guard running else { return }
        var request = URLRequest(url: URL(string: "https://slack.com/api/apps.connections.open")!)
        request.httpMethod = "POST"
        request.setValue("Bearer \(appToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        URLSession.shared.dataTask(with: request) { [weak self] data, _, _ in
            Task { @MainActor in
                guard let self, self.running else { return }
                guard let data, let obj = JSON.decode(data),
                      obj["ok"] as? Bool == true, let urlStr = obj["url"] as? String,
                      let url = URL(string: urlStr) else {
                    ChannelLog.write("slack: apps.connections.open failed — check the app-level token")
                    self.retry()
                    return
                }
                let t = URLSession.shared.webSocketTask(with: url)
                self.task = t
                t.resume()
                self.receive()
            }
        }.resume()
    }

    private func retry() {
        guard running else { return }
        DispatchQueue.main.asyncAfter(deadline: .now() + 5) { [weak self] in
            guard let self, self.running else { return }
            self.openSocket()
        }
    }

    private func receive() {
        task?.receive { [weak self] result in
            Task { @MainActor in
                guard let self, self.running else { return }
                switch result {
                case .failure:
                    self.task = nil
                    self.openSocket() // socket URLs are short-lived; reopen
                case .success(let message):
                    switch message {
                    case .string(let s): self.handleFrame(s)
                    case .data(let d): if let s = String(data: d, encoding: .utf8) { self.handleFrame(s) }
                    @unknown default: break
                    }
                    self.receive()
                }
            }
        }
    }

    private func handleFrame(_ raw: String) {
        guard let obj = JSON.decode(raw) else { return }
        let type = obj["type"] as? String

        // Every events_api / interactive / slash envelope must be acked by id.
        if let envelopeId = obj["envelope_id"] as? String {
            send(["envelope_id": envelopeId])
        }

        switch type {
        case "hello", "disconnect":
            if type == "disconnect" { task = nil; openSocket() }
        case "events_api":
            guard let payload = obj["payload"] as? [String: Any],
                  let event = payload["event"] as? [String: Any] else { return }
            onEvent(event)
        default:
            break
        }
    }

    private func onEvent(_ event: [String: Any]) {
        guard event["type"] as? String == "message" else { return }
        // Ignore bot messages, edits, and subtype events (joins, etc.).
        if event["bot_id"] != nil { return }
        if event["subtype"] != nil { return }
        let userId = event["user"] as? String ?? ""
        let text = (event["text"] as? String ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        let channel = event["channel"] as? String ?? ""
        guard !userId.isEmpty, !text.isEmpty, !channel.isEmpty else { return }

        let allow = MaxConfig.load().slackAllowlist
        guard allow.contains(userId) else {
            ChannelLog.write("slack: ignored message from user \(userId) — not in allowlist")
            return
        }

        let botToken = self.botToken
        ChannelRouter.shared.handle(key: "slack-\(channel)", text: text) { reply in
            Self.sendReply(channel: channel, text: reply, token: botToken)
        }
    }

    private func send(_ obj: [String: Any]) {
        task?.send(.string(JSON.encodeString(obj))) { _ in }
    }

    nonisolated static func sendReply(channel: String, text: String, token: String) {
        var request = URLRequest(url: URL(string: "https://slack.com/api/chat.postMessage")!)
        request.httpMethod = "POST"
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = JSON.encode(["channel": channel, "text": String(text.prefix(3500))])
        URLSession.shared.dataTask(with: request).resume()
    }
}
